### PR TITLE
Remove workaround for NativeAOT

### LINKED ISSUE #20: add a separate dir to build the NativeAoT version
some folks have expressed the desire to use this tool without having to install .NET 6. the .net 6 requirement is just for the NativeAoT which you can always build separately. so we should not require folks to have to have 6.0. 

my suggestion is we should have a separate dir that builds AoT that does require 6.0 (instead of having everyone who wants to use this and doesn't want to install 6.0 to change `src/GCRealTimeMon/GCRealTimeMon.csproj` themselves). the rest can use 5.0 as a min version (actually we could even make it 3.1...).

### PR DESCRIPTION
**Pull request by hez2010**
Since https://github.com/dotnet/runtimelab/pull/1728 fixed the issue where some metadata is missing after compilation, I removed the workaround in `rd.xml`.

Also fixes #20

### CODE REVIEW

**Comment by hez2010** [comment]
This PR should be held until the new ILCompiler build being uploaded to package source. Will remove the `DO NOT MERGE` in title once it's ready.

**Comment by hez2010** [comment]
This PR is now ready for merge.

**Comment by Maoni0** [comment]
hi @hez2010, so the resulting binary doesn't seem to work with the latest src, with or without this change... this is what I got

```
C:\realmon>C:\realmon\src\GCRealTimeMon\bin\Release\net6.0\win-x64\publish\gcrealtimemon -p 16128
Value cannot be null. Parameter name: path1
```
can you please try and see if this is what you get? I also got this build error -

```
C:\Program Files\dotnet\sdk\6.0.100-rc.2.21505.57\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.PackTool.targets(86,5): error NETSDK1053: Pack as tool does not support self contained. [C:\realmon\src\dotnet-gcmon\dotnet-gcmon.csproj]
```
looks like we shouldn't be trying to AoT this one or C:\realmon\tests\GCRealTimeMon.UnitTests\bin\Release\net6.0\win-x64\GCRealTimeMon.UnitTests.dll.

**Comment by hez2010** [comment]
Will investigate. Seems that some commits break the NativeAOT compilation so that I need to add more runtime directives.

**Comment by hez2010** [comment]
@Maoni0 Done.
I've also added support for .NET 5.0 and .NET Core 3.1.

File: src/GCRealTimeMon/Program.cs
Comment (Maoni0) [inline src/GCRealTimeMon/Program.cs]: nit - change to 

```csharp
                var location = Assembly.GetExecutingAssembly().Location;
                // single file mode doesn't support getting the location of an assembly
                if (string.IsNullOrEmpty(location)) 
                    configurationFile = "DefaultConfig.yaml";
                else
                    configurationFile = Path.Combine(Path.GetDirectoryName(location), "DefaultConfig.yaml");
```

**Review by Maoni0 — CHANGES_REQUESTED** [review]
also could you please do the change in `tests/GCRealTimeMon.UnitTests/GCRealTimeMon.UnitTests.csproj` as well?

the rest LGTM, thanks!

File: src/dotnet-gcmon/dotnet-gcmon.csproj
Comment (chrisnas) [inline src/dotnet-gcmon/dotnet-gcmon.csproj]: Since it is working with .NET Core 3.1, we should only build with netcoreapp3.1: it will work on 3.1, 5.0, and 6.0 and not only 6.0.
BTW, this is what is done for dotnet-counters (https://github.com/dotnet/diagnostics/blob/main/src/Tools/dotnet-counters/dotnet-counters.csproj#L4)

In addition, please update the \<PackageVersion\> to 0.3.0 to ensure that nuget will see the new package associated to a higher version as the currently uploaded.

**Review by chrisnas — CHANGES_REQUESTED** [review]

File: src/dotnet-gcmon/dotnet-gcmon.csproj
Comment (Maoni0) [reply src/dotnet-gcmon/dotnet-gcmon.csproj]: agree with @chrisnas - I noticed that it was building all 3 versions which seems unnecessary.

**Comment by Maoni0** [comment]
BTW, it would also be good to have some automatic testing for this because the AoTed version probably gets used very little. I don't use it myself and it looks like most people use the tool via the dotnet tools route.

File: src/GCRealTimeMon/Program.cs
Comment (hez2010) [reply src/GCRealTimeMon/Program.cs]: Done

File: src/dotnet-gcmon/dotnet-gcmon.csproj
Comment (hez2010) [reply src/dotnet-gcmon/dotnet-gcmon.csproj]: Done.

**Comment by hez2010** [comment]
I think we need to keep `net6.0` target to satisfy CI. Both `netcoreapp3.1` and `net5.0` are about to reach EOL. And seems that one can't run netcoreapp3.1 build of dotnet-gcmon if there's no .NET Core 3.1 runtime installed on the machine despite .NET 6 is present.

**Comment by chrisnas** [comment]
> I think we need to keep `net6.0` target to satisfy CI. Both `netcoreapp3.1` and `net5.0` are about to reach EOL. And seems that one can't run netcoreapp3.1 build of dotnet-gcmon if there's no .NET Core 3.1 runtime installed on the machine despite .NET 6 is present.

I don't know how dotnet tool install -g knows which "version" to extract from the nuget package...
The EOL is a Microsoft concept that is not "always" taken into account on customer sites (believe me ;^) and 3.1 should be kept for much longer. 
The only possible issue (i.e. having 3.1 only) that crossed my mind was forward compatibility; i.e. 3.1 gcmon trying to monitor a .NET 6 application. I tested it and it works (basically because the tool is run by the CLR 6)

**Comment by hez2010** [comment]
> I don't know how dotnet tool install -g knows which "version" to extract from the nuget package...

dotnet tool install doesn't know the exact runtime version from nuget. There's a .runtimeconfig.json present in the tool directory to indicate the proper runtime needed by the tool, and dotnet cli will choose a proper one to execute. 

So most tools are multi-targeted and they organize their content as following:

```
| net6.0
  \ - tool.dll
| net5.0
  \ - tool.dll
| netcoreapp3.1
  \ - tool.dll
.....
```

If there's no proper version of dotnet runtime installed on the machine, trying to run the tool will fail.

**Comment by chrisnas** [comment]
> > I don't know how dotnet tool install -g knows which "version" to extract from the nuget package...
> 
> dotnet tool install doesn't know the exact runtime version from nuget. There's a .runtimeconfig.json present in the tool directory to indicate the proper runtime needed by the tool, and dotnet cli will choose a proper one to execute.
> 
> So most tools are multi-targeted and they organize their content as following:
> 
> ```
> | net6.0
>   \ - tool.dll
> | net5.0
>   \ - tool.dll
> | netcoreapp3.1
>   \ - tool.dll
> .....
> ```
> 
> If there's no proper version of dotnet runtime installed on the machine, trying to run the tool will fail.

The .NET Diagnostics teams decided to keep only netcoreapp3.1 for all other CLI global tools.
image

I think we should do the same

**Comment by hez2010** [comment]
> I think we should do the same

I've downgraded the dotnet version in CI to 3.1. 

@Maoni0 PTAL

File: src/GCRealTimeMon/Program.cs
Comment (Maoni0) [reply src/GCRealTimeMon/Program.cs]: you don't need to get `Assembly.GetExecutingAssembly().Location` again, do you?

File: src/GCRealTimeMon/Program.cs
Comment (Maoni0) [inline src/GCRealTimeMon/Program.cs]: you don't need to get `Assembly.GetExecutingAssembly().Location` again, do you?

**Comment by Maoni0** [comment]
aside from that one minor comment (and conflicts need to be resolved), the rest LGTM!

File: src/dotnet-gcmon/dotnet-gcmon.csproj
Comment (chrisnas) [inline src/dotnet-gcmon/dotnet-gcmon.csproj]: Should be updated to 0.4.0 since another PR with 0.3.0 has been merged

**Review by chrisnas — CHANGES_REQUESTED** [review]

File: src/dotnet-gcmon/dotnet-gcmon.csproj
Comment (Maoni0) [reply src/dotnet-gcmon/dotnet-gcmon.csproj]: do we need to change this if I publish a new version with this change and the other change?

File: src/GCRealTimeMon/Program.cs
Comment (hez2010) [reply src/GCRealTimeMon/Program.cs]: Done.

**Comment by hez2010** [comment]
@Maoni0 Merge conflict has been resolved

File: src/dotnet-gcmon/dotnet-gcmon.csproj
Comment (chrisnas) [reply src/dotnet-gcmon/dotnet-gcmon.csproj]: > do we need to change this if I publish a new version with this change and the other change?

I thought you already published the version with the other PR.
No problem to release to nuget when all PRs are merged so you avoid bumping the version too often

**Comment by Maoni0** [comment]
thanks @hez2010!!

**Pull request merged**
